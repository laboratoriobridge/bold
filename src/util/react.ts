import { MutableRefObject, Ref } from 'react'

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}
