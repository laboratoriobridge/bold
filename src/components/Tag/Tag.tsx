import React, { CSSProperties } from 'react'

import { ExternalStyles, Theme, useStyles } from '../../styles'
import { Omit } from '../../util'
import { Icon, IconImage } from '../Icon'

export type TagType = 'normal' | 'danger' | 'info' | 'success' | 'alert'

export interface TagProps extends Omit<React.HTMLAttributes<HTMLSpanElement>, 'style'> {
  type?: TagType
  style?: ExternalStyles
  icon?: IconImage
  height?: String
  removable?: boolean
  onRemove?(): void
}

export function Tag(props: TagProps) {
  const { type, style, children, icon, height, removable, onRemove, ...rest } = props
  const { classes, css } = useStyles((theme) => createStyles(height, theme))

  return (
    <span className={css(classes.tag, classes[type], style)} {...rest}>
      {icon && <Icon icon={icon} style={classes.icon} />}
      {children}
      {removable && <Icon icon='timesDefault' style={classes.removeIcon} onClick={onRemove} />}
    </span>
  )
}

Tag.defaultProps = {
  height: '1.5rem',
  type: 'normal',
  onRemove: () => null,
} as Partial<TagProps>

const createStyles = (height: String, theme: Theme) => ({
  tag: {
    display: 'inline-flex',
    alignItems: 'center',
    padding: '0.25rem 0.5rem',
    fontWeight: 'bold',
    borderRadius: theme.radius.tag,
    whiteSpace: 'nowrap',
    lineHeight: '1rem',
    fontSize: theme.typography.sizes.text,
    height: height,
  } as CSSProperties,
  icon: {
    fontSize: '1rem',
    marginRight: '0.25rem',
  } as CSSProperties,
  removeIcon: {
    fontSize: '1rem',
    marginLeft: '0.125rem',
    '&:hover': {
      cursor: 'pointer',
      color: theme.pallete.status.danger.main,
    },
  },
  normal: {
    background: theme.pallete.gray.c80,
  } as CSSProperties,
  danger: {
    background: theme.pallete.status.danger.main,
    color: theme.pallete.status.danger.onColor,
  } as CSSProperties,
  info: {
    background: theme.pallete.status.info.main,
    color: theme.pallete.status.info.onColor,
  } as CSSProperties,
  success: {
    background: theme.pallete.status.success.main,
    color: theme.pallete.status.success.onColor,
  } as CSSProperties,
  alert: {
    background: theme.pallete.status.alert.main,
    color: theme.pallete.status.alert.onColor,
  } as CSSProperties,
})
