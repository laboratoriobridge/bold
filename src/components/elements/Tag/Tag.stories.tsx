import { storiesOf } from '@storybook/react'
import React from 'react'

import { HFlow } from '../Flow'

import { Tag } from './Tag'

storiesOf('Components|Tag', module).add('default', () => (
  <HFlow>
    <Tag>Normal</Tag>
    <Tag type='alert'>Alert</Tag>
    <Tag type='danger'>Danger</Tag>
    <Tag type='info'>Info</Tag>
    <Tag type='success'>Success</Tag>
  </HFlow>
))
