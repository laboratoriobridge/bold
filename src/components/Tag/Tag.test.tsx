import { fireEvent, render } from '@testing-library/react'
import React from 'react'

import { Tag } from './Tag'

it('should render correctly', () => {
  expect(render(<Tag>Normal</Tag>).container).toMatchSnapshot()
  expect(render(<Tag type='alert'>Alert</Tag>).container).toMatchSnapshot()
  expect(render(<Tag type='danger'>Danger</Tag>).container).toMatchSnapshot()
  expect(render(<Tag type='info'>Info</Tag>).container).toMatchSnapshot()
  expect(render(<Tag type='success'>Success</Tag>).container).toMatchSnapshot()
})

it('should accept the style prop', () => {
  expect(render(<Tag style={{ color: 'green' }}>Normal</Tag>).container).toMatchSnapshot()
})

it('should accept HTML span element properties', () => {
  const { container } = render(<Tag id='test'>Normal</Tag>)
  expect(container.querySelector('span')?.getAttribute('id')).toEqual('test')
})

it('should accept "icon" prop', () => {
  const { container } = render(<Tag icon='userFilled'>With icon</Tag>)
  expect(container).toMatchSnapshot()
})

it('should accept "height" prop and render with custom height', () => {
  const { container } = render(<Tag height={'2.0rem'}>Custom height</Tag>)
  expect(container).toMatchSnapshot()
})

it('should accept "removable" prop and render remove icon', () => {
  const { container } = render(<Tag removable>Removable</Tag>)
  expect(container).toMatchSnapshot()
})

it('should call onRemove when remove icon is clicked', () => {
  const remove = jest.fn()
  const { container } = render(
    <Tag removable onRemove={remove}>
      Removable
    </Tag>
  )
  expect(remove).not.toHaveBeenCalled()
  fireEvent.click(container.querySelector('svg'))
  expect(remove).toHaveBeenCalled()
})
