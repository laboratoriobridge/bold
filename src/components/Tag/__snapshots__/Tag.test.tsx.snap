// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept "height" prop and render with custom height 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.25rem 0.5rem;
  font-weight: bold;
  border-radius: 12px;
  white-space: nowrap;
  line-height: 1rem;
  font-size: 0.8125rem;
  height: 2.0rem;
  background: #D3D4DD;
}

<div>
  <span
    class="emotion-0"
  >
    Custom height
  </span>
</div>
`;

exports[`should accept "icon" prop 1`] = `
.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.25rem 0.5rem;
  font-weight: bold;
  border-radius: 12px;
  white-space: nowrap;
  line-height: 1rem;
  font-size: 0.8125rem;
  height: 1.5rem;
  background: #D3D4DD;
}

.emotion-0 {
  fill: currentColor;
  font-size: 1.5rem;
  font-size: 1rem;
  margin-right: 0.25rem;
}

<div>
  <span
    class="emotion-1"
  >
    <svg
      aria-hidden="true"
      class="emotion-0"
      height="1em"
      role="img"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M12 12a5 5 0 110-10 5 5 0 010 10zM2 20c0-4 2.001-6 6.004-6H16c4 0 6 2 6 6 0 1.333-.667 2-2 2H4c-1.333 0-2-.667-2-2z"
      />
    </svg>
    With icon
  </span>
</div>
`;

exports[`should accept "removable" prop and render remove icon 1`] = `
.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.25rem 0.5rem;
  font-weight: bold;
  border-radius: 12px;
  white-space: nowrap;
  line-height: 1rem;
  font-size: 0.8125rem;
  height: 1.5rem;
  background: #D3D4DD;
}

.emotion-0 {
  fill: currentColor;
  font-size: 1.5rem;
  font-size: 1rem;
  margin-left: 0.125rem;
}

.emotion-0:hover {
  cursor: pointer;
  color: #D01E29;
}

<div>
  <span
    class="emotion-1"
  >
    Removable
    <svg
      aria-hidden="true"
      class="emotion-0"
      height="1em"
      role="img"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        clip-rule="evenodd"
        d="M8.071 6.657l4.243 4.242 4.242-4.242c.464-.48.935-.48 1.415 0s.48.95 0 1.414l-4.243 4.243 4.243 4.242c.471.472.471.943 0 1.415-.472.471-.943.471-1.415 0l-4.242-4.243L8.07 17.97c-.46.482-.932.482-1.414 0-.482-.483-.482-.954 0-1.415l4.242-4.242L6.657 8.07c-.472-.47-.472-.942 0-1.414s.943-.472 1.414 0v.001z"
      />
    </svg>
  </span>
</div>
`;

exports[`should accept the style prop 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.25rem 0.5rem;
  font-weight: bold;
  border-radius: 12px;
  white-space: nowrap;
  line-height: 1rem;
  font-size: 0.8125rem;
  height: 1.5rem;
  background: #D3D4DD;
  color: green;
}

<div>
  <span
    class="emotion-0"
  >
    Normal
  </span>
</div>
`;

exports[`should render correctly 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.25rem 0.5rem;
  font-weight: bold;
  border-radius: 12px;
  white-space: nowrap;
  line-height: 1rem;
  font-size: 0.8125rem;
  height: 1.5rem;
  background: #D3D4DD;
}

<div>
  <span
    class="emotion-0"
  >
    Normal
  </span>
</div>
`;

exports[`should render correctly 2`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.25rem 0.5rem;
  font-weight: bold;
  border-radius: 12px;
  white-space: nowrap;
  line-height: 1rem;
  font-size: 0.8125rem;
  height: 1.5rem;
  background: #AD5000;
  color: #FFFFFF;
}

<div>
  <span
    class="emotion-0"
  >
    Alert
  </span>
</div>
`;

exports[`should render correctly 3`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.25rem 0.5rem;
  font-weight: bold;
  border-radius: 12px;
  white-space: nowrap;
  line-height: 1rem;
  font-size: 0.8125rem;
  height: 1.5rem;
  background: #D01E29;
  color: #FFFFFF;
}

<div>
  <span
    class="emotion-0"
  >
    Danger
  </span>
</div>
`;

exports[`should render correctly 4`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.25rem 0.5rem;
  font-weight: bold;
  border-radius: 12px;
  white-space: nowrap;
  line-height: 1rem;
  font-size: 0.8125rem;
  height: 1.5rem;
  background: #0069D0;
  color: #FFFFFF;
}

<div>
  <span
    class="emotion-0"
  >
    Info
  </span>
</div>
`;

exports[`should render correctly 5`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.25rem 0.5rem;
  font-weight: bold;
  border-radius: 12px;
  white-space: nowrap;
  line-height: 1rem;
  font-size: 0.8125rem;
  height: 1.5rem;
  background: #217B00;
  color: #FFFFFF;
}

<div>
  <span
    class="emotion-0"
  >
    Success
  </span>
</div>
`;
